src/server/frontend_xwayland/xwayland_wm_surface.cpp

```diff
@@ -118,46 +219,76 @@ void mf::XWaylandWMSurface::set_workspace(int workspace)
     xcb_flush(xwm->get_xcb_connection());
 }
 
-void mf::XWaylandWMSurface::set_wm_state(WmState state)
+void mf::XWaylandWMSurface::apply_mir_state_to_window(MirWindowState new_state)
 {
-    uint32_t const properties[]{
-        state,
-        XCB_WINDOW_NONE // Icon window
-    };
-
-    xcb_change_property(
-        xwm->get_xcb_connection(),
-        XCB_PROP_MODE_REPLACE,
-        window, xwm->xcb_atom.wm_state,
-        xwm->xcb_atom.wm_state, 32,
-        length_of(properties), properties);
-    xcb_flush(xwm->get_xcb_connection());
-}
-
-void mf::XWaylandWMSurface::set_net_wm_state()
-{
-    std::vector<uint32_t> properties;
+    WindowState new_window_state;
 
     {
         std::lock_guard<std::mutex> lock{mutex};
 
-        if (fullscreen)
-            properties.push_back(xwm->xcb_atom.net_wm_state_fullscreen);
+        if (new_state == cached_mir_window_state)
+            return;
+
+        cached_mir_window_state = new_state;
+        new_window_state = window_state;
 
-        if (maximized)
+        switch (new_state)
         {
-            properties.push_back(xwm->xcb_atom.net_wm_state_maximized_horz);
-            properties.push_back(xwm->xcb_atom.net_wm_state_maximized_vert);
+        case mir_window_state_minimized:
+        case mir_window_state_hidden:
+            new_window_state.minimized = true;
+            // don't change new_window_state.maximized
+            // don't change new_window_state.fullscreen
+            break;
+
+        case mir_window_state_fullscreen:
+            new_window_state.minimized = false;
+            // don't change new_window_state.maximized
+            new_window_state.fullscreen = true;
+            break;
+
+        case mir_window_state_maximized:
+        case mir_window_state_vertmaximized:
+        case mir_window_state_horizmaximized:
+        case mir_window_state_attached:
+            new_window_state.minimized = false;
+            new_window_state.maximized = true;
```

Comment:
I think you might need to split this into vert & horiz maximised state; there's a reasonable amount of stuff that does one but not the other?

Reply:
> there's a reasonable amount of stuff

Can't implement w/o an example I can test with